과제 1: 레스토랑 조회 구현하기

bbhye1 opened the pull request:
### **구현 내용**

TDD연습과 함께 Rest-API 적용도 천천히 진행할 예정입니다. 
중간 과정 중 부족한 부분이 있다면 코멘트 부탁드립니다! 😄

bbhye1 committed:
[red] Add dispatch test for App.

bbhye1 committed:
[red] Create test for App.

bbhye1 committed:
[green] Implement Restaurants text in App.

bbhye1 committed:
[red] Add test for App.

bbhye1 committed:
[green] Implement "서울" button in App.

bbhye1 committed:
[refactor] Create RestaurantRegions and apply to App.

bbhye1 committed:
[red] Create test for RestaurantRegions.

bbhye1 committed:
[green] Implement buttons in RestautantRegions.

bbhye1 committed:
[refactor] Refatoring buttons in RestaurantRegions.

bbhye1 committed:
[refactor] Create RestaurantRegionsContainer.

bbhye1 committed:
[refator] Create restaurantRegions fixture.

bbhye1 committed:
[red] Create useSelector test for RestaurantRegionsContainer.

bbhye1 committed:
[green] Create mock function.

bbhye1 committed:
[refactor] Create store and apply it.

bbhye1 committed:
[refactor] Add useSelector to App test.

bbhye1 committed:
[refactor] Add mock function to RestaurantRegions test.

bbhye1 committed:
[refactor] Change tags in RestaurantRegions.

bbhye1 committed:
[green] Implement dispatch in App.

bbhye1 committed:
[refactor] Seperate reducer form store.

bbhye1 committed:
[refactor] Fix test names.

bbhye1 committed:
[red] Create setRestaurantRegions test for reducer.

bbhye1 committed:
[green] Implement reducer.

bbhye1 committed:
[refactor] Add setRestaurantRegions in reducer.

bbhye1 committed:
[refactor] Create setRestaurantRegions action in actions.

bbhye1 committed:
[refactor] Seperate loadRestaurantRegions function from App.

bbhye1 committed:
[refactor] Rename from RestaurantRegions to Regions.

bbhye1 committed:
[refactor] Rename from RestaurantRegionsContainer to RegionsContainer.

bbhye1 committed:
[refactor] Rename fixture.

bbhye1 committed:
[refactor] rename All.

Kihyun92 commented on src/App.jsx:
```diff
@@ -1,7 +1,27 @@
 export default function App() {
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    loadRegions({ dispatch });
+  }, {});
```
```suggestion
  useEffect(() => {
    loadRegions({ dispatch });
  }, []);
```
최초 한번만 dispatch 하려는 지금과 같은 상황에는 두번째 인자로 빈배열을 넘겨줘야합니다.
- https://ko.reactjs.org/docs/hooks-reference.html#conditionally-firing-an-effect

Kihyun92 commented on src/Regions.jsx:
```diff
@@ -0,0 +1,18 @@
+      { regions.map(({ id, name }) => (
+        <li>
+          <button
+            key={id + name}
```
id의 경우 타입이 `number`이고, name은 `string`이죠. 이 코드는 의도가 명확하게 드러나지 않는 것 같습니다. 개인적으론 추후 `Typescript`를 사용할 때를 생각해서도 좋지 못한 방식인 것 같구요.

단순히 중복되지 않는 string 형식의 key 값을 세팅해주려는 목적이라면 `template literal`을 사용하는게 좋을 것 같습니다~
- https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Template_literals

Kihyun92 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,20 @@
+describe('reducer', () => {
+  describe('setRegions', () => {
+    it('changes regions', () => {
+      const initialRegions = [];
```
it 함수 밖에 선언해도 될 것 같아요~

Kihyun92 reviewed (changes requested):
일단 `Red-Green-Refactoring`의 사이클을 너무 잘 이용하시는 것 같아요!👍 그리고 `forEach`로 `side effect` 없는 테스트를 해보는것도 좋구요 ㅎㅎ 커멘트 드릴게 너무 없어서 걱정이 될정도네요 😂

bbhye1 committed:
[refactor] Apply setRegions action to reducer test.

bbhye1 committed:
[refactor] fix useEffect in App.

bbhye1 committed:
[refactor] Change key of buttons in Regions.

bbhye1 committed:
[red] Add undefined action test for reducer.

bbhye1 committed:
[green] Implement no action situation.

bbhye1 committed:
[refactor] Set the initialState in reducer.

bbhye1 committed:
[refactor] Set the initialAction in reducer.

bbhye1 committed:
[refactor] Create fetchRegions to fetch Regions data.

bbhye1 committed:
[refactor] Fetch regions data.

bbhye1 committed:
[refactor] Create mock api functions

bbhye1 committed:
[refactor] Apply thunk in store.

bbhye1 committed:
[refactor] Move loadRegions function to actions to apply thunk.

bbhye1 committed:
[refactor] Move key props from button to li.

bbhye1 committed:
[refactor] Correct the wrong url in fetchRegions.

bbhye1 committed:
[refactor] Refactoring App test.

bbhye1 committed:
[red] Add test with categories fixture for App.

bbhye1 committed:
[green] Implement buttons in App.

bbhye1 committed:
[refactor] Create Categories.

bbhye1 committed:
[red] Create test for Categories.

bbhye1 committed:
[green] Implement Categories.

bbhye1 committed:
[refactor] refactoring Categories.

bbhye1 committed:
[refactor] Create CategoriesContainer for Categories.

bbhye1 committed:
[red] Create test for CategoriesContainer.

bbhye1 committed:
[green] Implement CategoriesContainer.

bbhye1 committed:
[refactor] Apply useSelector to CategoriesContainer.

bbhye1 committed:
[refactor] Create loadCategories.

bbhye1 committed:
[refactor] Create fetchCategories to fetch categories data.

bbhye1 committed:
[refactor] Add setCategories action.

bbhye1 committed:
[red] Add setCategories test for reducer.

bbhye1 committed:
[refactor] Implement setCategories in reducers.

bbhye1 committed:
[refactor] Refactoring setCategories in reducers.

bbhye1 committed:
[refactor] Remove duplication in reducers.

bbhye1 committed:
[refactor] Change dispatch test to make it more meaningful.

bbhye1 committed:
[refactor] Change the position of lines.

bbhye1 committed:
[red] Add click event test for Categories.

bbhye1 committed:
[green] Implement onClick in Categories.

bbhye1 committed:
[red] Add dispatch test for Categories.

bbhye1 committed:
[green] Implement dispatch in CategoriesContainer.

bbhye1 committed:
[refactor] Create selectCategory action.

bbhye1 committed:
[refactor] Change the matcher in App test.

Change the matcher from toBeCalledTimes(2) to toBeCalled() because more dispatch are added.

bbhye1 committed:
[red] Add checking dispatch action test for CategoriesContainer.

bbhye1 committed:
[green] Set the parameter in Categories & CategoriesContainer.

bbhye1 committed:
[red] Add selectCategory test for reducer.

bbhye1 committed:
[green] Implement selectCategory in reducer.

bbhye1 committed:
[refactor] refactoring selectCategory in reducer.

bbhye1 committed:
[refactor] Refactoring reducer test

bbhye1 committed:
[refactor] Fix onClick function in Categories.

bbhye1 committed:
[red] Add another selectCategory test for reducer.

bbhye1 committed:
[green] Implement selectCategory in reducer.

bbhye1 committed:
[refactor] Refactoring selectCategory in reducers.

bbhye1 committed:
[red] Add click event test for Regions.

bbhye1 committed:
[green] Implement onClick in Regions.

bbhye1 committed:
[red] Add dispatch test for Regions.

bbhye1 committed:
[green] Implement dispatch in RegionsContainer.

bbhye1 committed:
[refactor] Create selectRegion action.

bbhye1 committed:
[red] Add checking dispatch action test for RegionsContainer.

bbhye1 committed:
[green] Put the callback function into onClick in Regions.

bbhye1 committed:
[red] Add selectRegion test for reducer.

bbhye1 committed:
[green] Implement selectRegion in reducers.

bbhye1 committed:
[refactor] Refactoring selectRegion in reducer.

bbhye1 committed:
[red] Add another selectRegion test for reducer.

bbhye1 committed:
[green] Implement selectRegion in reducer.

bbhye1 committed:
[refactor] Refactoring selectRegion in reducers.

bbhye1 committed:
[red] Add selected category test for reducer.

bbhye1 committed:
[green] Implement selected category in reducer.

bbhye1 committed:
[refactor] Refactoring selected category.

bbhye1 committed:
[red] Add selected region test for reducer.

bbhye1 committed:
[green] Implement selected region in reducer.

bbhye1 committed:
[refactor] Add name parameter in selectRegion action.

bbhye1 committed:
[refactor] refactoring All paramenter of selectedRegion action.

bbhye1 committed:
[refactor] Add selectedRegion to initialState in reducer.

bbhye1 commented:
카테고리와 지역을 클릭하면 카테고리와 지역이 저장되는 부분까지 구현했습니다. 
이제 레스토랑 리스트를 보여주는 작업이랑 마지막 전체 리팩토링이 남았네요!  내일까지 빠르게 구현하고 리팩토링에 집중하려고 합니다.

중간에  selectCategory를 reducer에 구현하는 부분에 조금 헤메서 커밋안하고 한번에 쭉했다가 다시 되돌아와 커밋시작했는데 그부분이 뒤죽박죽 커밋이 됐네요 ㅋㅋㅋ😂 한참 후에 다시 보다가 발견해서 그냥 그대로 두었습니다!

bbhye1 committed:
[red] Add checking text test for App.

bbhye1 committed:
[green] Implement text in App.

bbhye1 committed:
[refactor] Create Restaurants component.

bbhye1 committed:
[red] Create test for Restaurants.

bbhye1 committed:
[green] Implement Restaurants.

bbhye1 committed:
[refactor] Refactoring Restautants.

bbhye1 committed:
[refactor] Remove useless test in App.

bbhye1 committed:
[refactor] Create RestaurantsContainer.

bbhye1 committed:
[refactor] Changes the name of tests in reducer.

bbhye1 committed:
[red] Add test for App.

bbhye1 committed:
[green] Implement App.

bbhye1 committed:
[refactor] Create restaurants fixture.

bbhye1 committed:
[refactor] Remove hardcoding in App.

bbhye1 committed:
[red] Create test for RestaurantsContainer.

bbhye1 committed:
[green] Implement RestaurantsContainer.

bbhye1 committed:
[refactor] Add restaurants to initialState in reducer.

Kihyun92 commented on src/Categories.jsx:
```diff
@@ -0,0 +1,20 @@
+    <ul>
+      {
+        categories.map(({ id, name }) => (
+          <li key={name}>
```
key 속성값을 name으로 한 이유가 있나요? name이 중복되지 않는다는 보장이 있다면 문제없지만, id로 하는게 더 좋을 것 같아요~

bbhye1 replied:
아래의 regions 에서의 id도 이고 categories의 id들도 숫자라서 혹시 겹치는것 아닌가 했어요. 각 컴포넌트안에서만 고유하면 괜찮나요??

Kihyun92 replied:
각 배열 내에서만 고유하면 됩니다~

이 문서 참고해보시면 좋겠네요 :)
- https://ko.reactjs.org/docs/lists-and-keys.html#keys-must-only-be-unique-among-siblings

Kihyun92 commented on src/RegionsContainer.jsx:
```diff
@@ -0,0 +1,26 @@
+    regions: state.regions,
+  }));
+
+  function handleClickSelectRegion(selectedRegion) {
```
`handleClickRegion`은 어떤가요? click과 select가 비슷한 의미를 갖는 것 같아서요~

Kihyun92 commented on src/reducer.js:
```diff
@@ -0,0 +1,69 @@
+
+        return {
+          ...category,
+          name: id === payload.id
+            ? `${name.replace('(v)', '')}(v)`
+            : name.replace('(v)', ''),
```
이곳에서 name에 체크 표시를 해주려다 보니 조금 복잡해보여요. 체크 표시 노출 여부는 `presentation component`에게 위임하고, 이곳에서는 체크 여부를 내려주는건 어떨까요?🤔

Kihyun92 reviewed (changes requested):
매우 좋습니다~ `red-green-refactoring` 사이클에 속도가 점점 붙으시는 것 같아요 :) 💯

bbhye1 committed:
[green] Create loadRegions test for actions.

bbhye1 committed:
[green] Add loadCategories test for actions.

bbhye1 committed:
[red] Add dispatch test for CategoriesContainer.

bbhye1 committed:
[green] Implement dispatch in CategoriesContainer.

bbhye1 committed:
[refactor] Create setRestaurants action.

bbhye1 committed:
[refactor] Create loadRestaurants action.

bbhye1 committed:
[red] Add loadRestaurants test for actions.

bbhye1 committed:
[green] Implement loadRestaurants in actions.

bbhye1 committed:
[refactor] Create fetchRestaurants in api.

bbhye1 committed:
[refactor] Add mock value to test in actions.

bbhye1 committed:
[refactor] Change the function name in Categories test

bbhye1 committed:
[refactor] Change the function name in CategoriesContainer.

bbhye1 committed:
[refactor] Change the function name in Regions test.

bbhye1 committed:
[refactor] Change the function name in RegionsContainer.

bbhye1 committed:
[refactor] Change the key of Categoires and Regions.

bbhye1 committed:
[red] Change the selectCategory test for reducer.

bbhye1 committed:
[green] Change the selectCategory in reducer.

bbhye1 committed:
[red] Change selectRegion test for reducer.

bbhye1 committed:
[gree] Change selectRegion in reducer.

bbhye1 committed:
[refactor] remove checked and checked test in reducer.

bbhye1 committed:
[refactor] Refactoring test for Categories

bbhye1 committed:
[red] Add button text test for Categories.

bbhye1 committed:
[green] Implement checked in Categories.

bbhye1 committed:
[refactor] Refactoring Regions test.

bbhye1 committed:
[red] Add button test for Regions.

bbhye1 committed:
[green] Implement checked in Regions.

bbhye1 committed:
[refactor] Add selectedRegion to RegionsContainer.

bbhye1 committed:
[refactor] Add selectedCategory in CategoriesContainer.

bbhye1 committed:
[refactor] Refactoring CategoriesContainer test.

bbhye1 committed:
[refactor] Add fetch to fetchRestaurants in api.

bbhye1 committed:
[refactor] Dispatch loadRestaurants in CategoriesContainer.

bbhye1 committed:
[refactor] Change the name of parameter in fetchRestaurants.

bbhye1 committed:
[refactor] Change loadRestaurants test in action.

bbhye1 committed:
[refactor] Get state at loadRestaurants in actions.

bbhye1 committed:
[refactor] Reorder the position of dispatches in CategoriesContainer.

bbhye1 commented:
일단 지역을 누르고 카테고리를 누르면 레스토랑 정보를 조회해서 store에 저장하는 것까지는 진행했는데 저장된 정보가 비동기로 저장되어서 RestaurantsContainer가 먼저 랜더가 되어서 정보가 화면에 보여지지 않네요 ㅜㅜ;; 이건 어떻게 해결 할 수 있을까요?

뒤로갈 수록 점점 헷갈려서  TDD패턴도 잘 정리가 안되네요 😂 😂

bbhye1 committed:
[refactor] Refactoring Categories test .

bbhye1 committed:
[refactor] Refactoring CategoriesContainer test.

bbhye1 committed:
[refactor] Refactoring Regions test

bbhye1 committed:
[refactor] Refactoring RegionsContainer test

bbhye1 committed:
Install jest-fetch-mock.

bbhye1 committed:
[green] Create fetchRegions test for api.

bbhye1 committed:
[green] Add fetchCategories test for api.

bbhye1 committed:
[green] Add fetchRestaurants test for api.

bbhye1 committed:
[refactor] Change all text of check.

bbhye1 committed:
[red] Add setRestaurants test for reducer.

bbhye1 committed:
[green] Implement setRestaurants in reducer.

bbhye1 committed:
[refactor] Refactoring setRestaurants in reducer.

bbhye1 committed:
[red] Add another test for loadRestaurants in actions..

bbhye1 committed:
[green] Implement loadRestaurants in actions.

bbhye1 committed:
[refactor] Refactoring loadREstaurants in actions.

bbhye1 committed:
[refactor] dispatch loadRestaurants in RegionsContainer.

bbhye1 committed:
[refactor] Refactoring Categories test.

bbhye1 committed:
[refactor] Refactoring CategoriesContainer test.

bbhye1 committed:
[refactor] Refactoreing test for reducer.

bbhye1 committed:
[refactor] Refactoring test for  RegionsContainer.

bbhye1 committed:
[refactor] Refactoring test for Restaurants.

bbhye1 committed:
[refactor] Refactoring test for RestaurantsContainer.

bbhye1 committed:
[refactor] Refactoring test for Restaurants.

bbhye1 committed:
[refactor] Refactoring Categories.

Kihyun92 commented:
소스 받아서 확인해보니 잘 노출 되는 것 같은데, @bbhye1 님 혹시 해결 하셨나요? 제가 확인이 늦었네요 🙏

bbhye1 commented:
> 소스 받아서 확인해보니 잘 노출 되는 것 같은데, @bbhye1 님 혹시 해결 하셨나요? 제가 확인이 늦었네요 🙏

네 ㅎㅎㅎ 리듀서에 액션을 안만들어주는 실수를 해놓고선  왜안될까만 고민하고 있었더라구요 ㅎㅎㅎㅎ 수정했습니다.

bbhye1 committed:
[refactor] Change describe to context in RegionsContainer test.

bbhye1 committed:
[refactor] Change describe to context in CategoriesContainer test.

bbhye1 commented:
중간에 api test는 테스트하는 법을 몰라서 그냥 쭉 구현했다가 나중에 찾아보고 작성해서 모두 통과하는 테스트라 쭉 green으로 올렸습니다. 테스트까지  완성하고 리팩토링 할 부분 조금씩 수정했습니다. 부족한 부분 리뷰 부탁드립니다!

bbhye1 committed:
[refactor] Reorder import lines.

Kihyun92 commented on src/CategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,55 @@
+    categories.forEach(({ name }) => {
+      expect(container).toHaveTextContent(name);
+
+      fireEvent.click(getByText(name));
```
이 테스트는 카테고리들이 잘 렌더링 되는지 테스트하는 곳인 것 같네요~ 
그런데 이곳에서 click을 처리해주는 이유가 있나요? 아래에서 따로 테스트 하고 있기도 하구요. 혹시 다른 이유가 있다면 말씀해주세요~ :)

bbhye1 replied:
없는게 맞습니다.  테스트를 분리하는 과정에서 잘 못보고 추가된것같아요ㅜㅜ 삭제했습니다!

Kihyun92 commented on src/reducer.js:
```diff
@@ -0,0 +1,45 @@
+  }),
+  selectRegion: (state, payload) => ({
+    ...state,
+    selectedRegion: payload.name,
```
payload에 `id, name` 두개의 값을 넘겨주는데, name만 사용하고 있는 것 같아요. 하나의 값으로 처리하는 방법은 없을까요?

예를 들어 id만 넘긴다면 state 내에 region 에서 찾아서 쓸수 있을 것 같네요. 객체로 넘길지라도 프로퍼티가 한개냐 두개냐에 따라 이해하는데 차이가 있을 것 같아요~

bbhye1 replied:
의미 면에서는 말씀해주신 내용대로 id값을 파라미터로 받아와서 state 내 region에서 찾고 그것의 name을 selectedRegion값에 넣어주는게 의미를 명시적이라 좋을거같지만 사실 이미 button 자체가 state값을 가져와 나열한것이기 때문에  굳이 다시 regions에서 찾아줘야 하는 번거로운 과정을 거쳐야 하나 싶기도 하거든요. button들이 랜더링 되었다는 자체가 해당 상태값이 존재한다는걸 증명해주니까요. 그래서 그냥 name을 바로 파라미터 값으로 받아오면 위에  selectCategory에서 id를 받아오는 것처럼 코드가 더 간결해질수 있다고 생각해서 변경했습니다. 이부분에 대해 의견이 궁금합니다! 😄

Kihyun92 replied:
말씀하신대로 name으로 처리해주는 것도 괜찮은 것 같네요~👍 

제가 생각했던 부분은 id 값만 전달해주더라도, id 값을 기반으로 state에서 name 정보를 찾을 수 있고, 개인적으로 고유한 id 값을 통해서 데이터들을 다루는 방식을 선호하기 때문에 말씀 드렸던 것 같아요 ㅎㅎ
제 피드백보다 더 좋은 고민을 해보신 것 같아 좋네요 ☺️

Kihyun92 commented on src/reducer.js:
```diff
@@ -0,0 +1,45 @@
+    ...state,
+    restaurants,
+  }),
+  selectCategory: (state, payload) => ({
```
```suggestion
  selectCategory: (state, {id}) => ({
```
다른 메서드 처럼 구조 분해 할당으로 처리할 수 있을 것 같네요~

Kihyun92 reviewed (changes requested):
리뷰가 많이 늦어 죄송합니다 😭 혜원님이 잘 진행해주셔서 피드백 드리기가 쉽지 않네요😅
테스트를 잘 작성해주고 계신 것 같아요~ 특히 API에 대한 테스트 해주신것에 대해서는 좋은 고민 해보신 것 같습니다~ 제 생각에도 프론트 단에서 API에 대해서 TDD로 작성하기 어려운 부분이 어느정도 존재할 것 같습니다. 아무래도 구현된 스펙에 따라서 연동이 이루어지니까요. 자연스레 작업하신것처럼 API 연동 이후에 성공하는 테스트 코드를 작성할 수 밖에 없을 것 같아요 :)

지금처럼 여러 시도와 고민해보시고 공유해주시면 좋겠습니다 👍

bbhye1 committed:
[refactor] Remove click event in CategoriesContainer test.

bbhye1 committed:
[refactor] Refactor selectRegion in reducer.

Change the way of find selectedRegion.

bbhye1 committed:
[red] Change the parameter of selectRegion in reducer test.

bbhye1 committed:
[green] Change the parameter of selectRegion in actions.

bbhye1 committed:
[refactor] Change parameters in Regions and RegionsCaontainer

bbhye1 committed:
[refator] Fix the test for RegionsContainer.

bbhye1 committed:
[refactor] Change the parameter in selectRegion action from id to name.

bbhye1 committed:
[refactor] Change test name in Categories test.

bbhye1 committed:
[refactor] Change test names in Regions test.

Kihyun92 reviewed (approved):
고생하셨습니다! 쉽지 않은 과제였는데, 잘해주셨어요 💯 
`Red-Green-Refactoring` 사이클도 잘 적용하시고, Redux와 API 연동 부분을 잘 수행해주셨어요. 내일 올라오는 과제 풀이 영상 꼭 보시면 좋을 것 같습니다~

이번 과제를 통해 어떤걸 배우셨는지, 그리고 더 해보고 싶은 부분이 있는지 고민해보시면 좋을 것 같아요~ 😁
혹시 더 고민을 나누고 싶은 부분이 있다면 말씀해주세요~
